### PR TITLE
[2.28] Rewrite x25519 example program

### PR DESCRIPTION
**Pull request by tom-daubney-arm**
Backport of #5725

### CODE REVIEW

File: programs/pkey/ecdh_curve25519.c [hunk 1]
```diff
@@ -118,105 +118,89 @@ int main( int argc, char *argv[] )
...
-    {
-        mbedtls_printf( " failed\n  ! mbedtls_mpi_read_binary returned %d\n", ret );
-        goto exit;
-    }
+    size_t secret_cli_olen = olen;
 
-    ret = mbedtls_ecdh_compute_shared( &ctx_cli.grp, &ctx_cli.z,
-                                       &ctx_cli.Qp, &ctx_cli.d,
-                                       mbedtls_ctr_drbg_random, &ctr_drbg );
+    ret = mbedtls_ecdh_calc_secret( &ctx_srv, &olen, secret_srv,
```
Comment (paul-elliott-arm) [inline programs/pkey/ecdh_curve25519.c:180]:
```suggestion
    ret = mbedtls_ecdh_calc_secret( &ctx_srv, &srv_olen, secret_srv,
```

File: programs/pkey/ecdh_curve25519.c [hunk 2]
```diff
@@ -118,105 +118,89 @@ int main( int argc, char *argv[] )
...
     /*
      * Verification: are the computed secrets equal?
      */
-    mbedtls_printf( "  . Checking if both computed secrets are equal..." );
+    mbedtls_printf( "  . Check if both calculated secrets are equal..." );
     fflush( stdout );
 
-    ret = mbedtls_mpi_cmp_mpi( &ctx_cli.z, &ctx_srv.z );
-    if( ret != 0 )
+    ret = memcmp( secret_srv, secret_cli, sizeof( secret_srv_olen ) );
```
Comment (paul-elliott-arm) [inline programs/pkey/ecdh_curve25519.c:200]:
```suggestion
    ret = memcmp( secret_srv, secret_cli, srv_olen );
```

File: programs/pkey/ecdh_curve25519.c [hunk 3]
```diff
@@ -118,105 +118,89 @@ int main( int argc, char *argv[] )
...
+    if( ret != 0 || ( secret_cli_olen != secret_srv_olen ) )
```
Comment (paul-elliott-arm) [inline programs/pkey/ecdh_curve25519.c:201]:
```suggestion
    if( ret != 0 || ( cli_olen != srv_olen ) )
```

**Review by paul-elliott-arm — CHANGES_REQUESTED** [review]
This still needs the Changelog entry. We build the Changelog for all LTS branches.

**Review by gilles-peskine-arm — APPROVED** [review]
I checked manually that `programs/pkey/ecdh_curve25519` now works whether Everest is enabled or not.

**Pull request merged**